From: John Crispin <blogic@openwrt.org>
Date: Thu, 8 Aug 2013 14:02:23 +0200

[PATCH 02/34] MIPS: lantiq: adds minimal dcdc driver

From 82a678435e9937dc8a7cb801f476e340fcfbc23e Mon Sep 17 00:00:00 2001
This driver so far only reads the core voltage.

Signed-off-by: John Crispin <blogic@openwrt.org>

Acked-by: John Crispin <blogic@openwrt.org>
Patchwork: http://patchwork.linux-mips.org/patch/5677/
---
 arch/mips/lantiq/xway/Makefile |    2 +-
 arch/mips/lantiq/xway/dcdc.c   |   63 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 64 insertions(+), 1 deletion(-)
 create mode 100644 arch/mips/lantiq/xway/dcdc.c

diff --git a/arch/mips/lantiq/xway/Makefile b/arch/mips/lantiq/xway/Makefile
--- a/arch/mips/lantiq/xway/Makefile
+++ b/arch/mips/lantiq/xway/Makefile
@@ -1,4 +1,4 @@
-obj-y := prom.o sysctrl.o clk.o reset.o ltq_dma.o timer.o mtlk_init.o
+obj-y := prom.o sysctrl.o clk.o reset.o ltq_dma.o timer.o mtlk_init.o dcdc.o
 
 obj-y += vmmc.o mtd_split.o
 
diff --git a/arch/mips/lantiq/xway/dcdc.c b/arch/mips/lantiq/xway/dcdc.c
new file mode 100644
--- /dev/null
+++ b/arch/mips/lantiq/xway/dcdc.c
@@ -0,0 +1,82 @@
+/*
+ *  This program is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 as published
+ *  by the Free Software Foundation.
+ *
+ *  Copyright (C) 2012 John Crispin <blogic@openwrt.org>
+ *  Copyright (C) 2010 Sameer Ahmad, Lantiq GmbH
+ */
+
+#include <linux/ioport.h>
+#include <linux/of_platform.h>
+
+#include <lantiq_soc.h>
+
+/* Bias and regulator Setup Register */
+#define DCDC_BIAS_VREG0	0xa
+/* Bias and regulator Setup Register */
+#define DCDC_BIAS_VREG1	0xb
+
+/* Interrupt Enable Register for TEMPSENSOR and DCDC2V5ERR */
+#define IRNEN		0x4e8
+
+#define dcdc_w8(x, y)		ltq_w8((x), dcdc_membase + (y))
+#define dcdc_r8(x)		ltq_r8(dcdc_membase + (x))
+#define dcdc_w32(x, y)		ltq_w32((x), dcdc_membase + (y))
+#define dcdc_r32(x)		ltq_r32(dcdc_membase + (x))
+
+static void __iomem *dcdc_membase;
+
+
+void ltq_xrx300_ena_tempsensor_irq(int ena)
+{
+	unsigned int val;
+	
+	val = dcdc_r32(IRNEN);
+	val &= ~0x2;
+	if(ena == 1){
+		val |= 0x2;
+	}
+	dcdc_w32(val, IRNEN);
+}
+EXPORT_SYMBOL_GPL(ltq_xrx300_ena_tempsensor_irq);
+
+static int dcdc_probe(struct platform_device *pdev)
+{
+	struct resource *res;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	dcdc_membase = devm_ioremap_resource(&pdev->dev, res);
+	if (IS_ERR(dcdc_membase))
+		return PTR_ERR(dcdc_membase);
+
+	dev_info(&pdev->dev, "Core Voltage : %d mV\n",
+		dcdc_r8(DCDC_BIAS_VREG1) * 8);
+
+	return 0;
+}
+
+static const struct of_device_id dcdc_match[] = {
+	{ .compatible = "lantiq,dcdc-xrx200" },
+	{},
+};
+
+static struct platform_driver dcdc_driver = {
+	.probe = dcdc_probe,
+	.driver = {
+		.name = "dcdc-xrx200",
+		.owner = THIS_MODULE,
+		.of_match_table = dcdc_match,
+	},
+};
+
+int __init dcdc_init(void)
+{
+	int ret = platform_driver_register(&dcdc_driver);
+
+	if (ret)
+		pr_info("dcdc: Error registering platform driver\n");
+	return ret;
+}
+
+arch_initcall(dcdc_init);
